### ISSUE TITLE
[CLOSED] Develop

### ISSUE THREAD

**Issue by schuhschuh** [cmake-basis/BASIS#232]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [ahundt](https://github.com/ahundt)**
_Monday Nov 11, 2013 at 00:40 GMT_
_Originally opened as https://github.com/schuhschuh/cmake-basis-legacy/pull/232_

---

This branch is for unreleased but completed features. I've merged response files into this branch.

---

_**[ahundt](https://github.com/ahundt)** included the following code: https://github.com/schuhschuh/cmake-basis-legacy/pull/232/commits_

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Nov 11, 2013 at 00:43 GMT_

---

Even though I didn't realize what a pull request meant, I don't see any harm in actually merging the response file updates into release.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Monday Nov 11, 2013 at 00:53 GMT_

---

My current idea of adding a custom configuration to the project template is a little bit more "Pythonic". Not sure if the response file will then still have a significant use. What do you think of a template configuration file such as the following Python code in `<template_dir>/config.py`. This also solves (yet only partially) the custom substitutions issue. The current draft reflects all options of `basisproject` regarding template selection.

``` python
# project template configuration script for basisproject tool

from os.path  import abspath
from datetime import datetime as date

# ------------------------------------------------------------------------------
template = {
  'version'   : 1.8,
  'directory' : abspath(__file__)
}

# ------------------------------------------------------------------------------
required = [
  # root documentation files
  'AUTHORS.txt',
  'README.txt',
  'INSTALL.txt',
  'COPYING.txt',
  'ChangeLog.txt',
  # root CMake configuration
  'CMakeLists.txt',
  'BasisProject.cmake'
]

# ------------------------------------------------------------------------------
options = [
  # additional configuration files
  {
    'name' : 'config',
    'desc' : 'Include/exclude common custom configuration files.',
    'path' : [
               'config/Settings.cmake',
               'config/Config.cmake.in',
               'config/ConfigSettings.cmake'
             ]
  },
  {
    'name' : 'config-settings',
    'desc' : 'Include/exclude custom Settings.cmake file.',
    'path' : 'config/Settings.cmake'
  },
  {
    'name' : 'config-components',
    'desc' : 'Include/exclude custom Components.cmake file.',
    'path' : 'config/Components.cmake'
  },
  {
    'name' : 'config-package',
    'desc' : 'Include/exclude custom Package.cmake file.',
    'path' : 'config/Package.cmake'
  },
  {
    'name' : 'config-find',
    'desc' : 'Include/exclude custom Config.cmake.in file.',
    'path' : [
               'config/Config.cmake.in',
               'config/ConfigSettings.cmake'
             ]
  },
  {
    'name' : 'config-find-version',
    'desc' : 'Include/exclude custom ConfigVersion.cmake.in file.',
    'path' : 'config/ConfigVersion.cmake.in'
  },
  {
    'name' : 'config-script',
    'desc' : 'Include/exclude custom ScriptConfig.cmake.in file.',
    'path' : 'config/ScriptConfig.cmake.in'
  },
  {
    'name' : 'config-test',
    'desc' : 'Include/exclude custom CTestCustom.cmake.in file.',
    'path' : 'config/CTestCustom.cmake.in'
  },
  {
    'name' : 'config-use',
    'desc' : 'Include/exclude custom ConfigUse.cmake.in file.',
    'path' : 'config/ConfigUse.cmake.in'
  },
  # software data
  {
    'name' : 'data',
    'desc' : 'Add/remove directory for auxiliary data files.',
    'path' : 'data/CMakeLists.txt'
  },
  # documentation
  {
    'name' : 'doc',
    'desc' : 'Add/remove directory for documentation files.',
    'path' : 'doc/CMakeLists.txt'
  },
  {
    'name' : 'doc-rst',
    'desc' : 'Add/remove reStructuredText (.rst) files for software manual/web site.',
    'path' : [
               'doc/CMakeLists.txt',
               'doc/index.rst',
               'doc/changelog.rst',
               'doc/download.rst',
               'doc/installation.rst',
               'doc/manual.rst',
               'doc/publications.rst',
               'doc/people.rst'
             ]
  },
  # usage example
  {
    'name' : 'example',
    'desc' : 'Add/remove directory for example files.',
    'path' : 'example/CMakeLists.txt'
  },
  # project modules
  {
    'name' : 'modules',
    'desc' : 'Add/remove support for modularization.',
    'path' : 'modules/'
  },
  # source files
  {
    'name' : 'include',
    'desc' : 'Add/remove directory for public header files.',
    'path' : 'include/'
  },
  {
    'name' : 'src',
    'desc' : 'Add/remove directory for project source files.',
    'path' : 'src/CMakeLists.txt'
  },
  # testing tree
  {
    'name' : 'test',
    'desc' : 'Add/remove support for testing.',
    'path' : [
               'CTestConfig.cmake',
               'test/CMakeLists.txt'
             ]
  },
  {
    'name' : 'test-internal',
    'desc' : 'Add/remove support for internal testing.',
    'path' : [
               'test/CMakeLists.txt',
               'test/internal/CMakeLists.txt'
             ]
  }
}

# ------------------------------------------------------------------------------
# preset templates
presets = {
  'minimal'  : [],
  'standard' : [ 'doc', 'doc-rst', 'example', 'noconfig', 'nodata', 'test' ],
  'toplevel' : [ 'modules', 'noinclude', 'nosrc' ],
  'module'   : [ 'nomodules', 'include', 'src', 'test' ],
  'full'     : [ 'config', 'doc', 'doc-rst', 'example', 'data', 'include', 'src', 'test' ]
}

# ------------------------------------------------------------------------------
substitutions = {
  'template-version'   : template['version'],
  'template-directory' : template['directory'],
  'project'            : project ['name'],
  'project_l'          : project ['name'].lower(),
  'description'        : project ['description'],
  'author'             : project ['author'],
  'date'               : date.today().strftime('%x'),
  'day'                : date.today().day,
  'month'              : date.today().month,
  'year'               : date.today().year
}

```

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Nov 11, 2013 at 01:00 GMT_

---

Well the reason I want response files isn't for project template configuration. It is for users running basisproject so I can send them a response file already set up for them, and they just have to fill out the parts they need. Even with a more advanced configuration file in the template I still see value in the response file capability.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Nov 11, 2013 at 01:07 GMT_

---

I'm not far enough along with my python understanding to fully understand precisely what the syntax is doing. Is it creating objects with the text on the left being fields and the text on the right being the value of the field? If so, that setup looks pretty good to me. Perhaps this part of the discussion belongs under a related issue rather than the pull request?

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Monday Nov 11, 2013 at 01:07 GMT_

---

Yes, well I understand and it won't hurt. Just thought first they will have to edit most parts of this file anyway, so why not just do this when typing the command ? Instead of sending them a response file, you send them just:

`basisproject --name "NameOfProject" --description "Brief project description." --standard`

Then they can even just copy and paste the command. When they later want to update the project, the initial response file would not work. It will only work once, so I have still a few doubts about the added value. All the common extra template options can be aggregated in a preset specified in the `config.py` of the tempate, such as `--standard` above.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Monday Nov 11, 2013 at 01:09 GMT_

---

Indeed, issue #231 would be the right place to continue the discussion.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Nov 11, 2013 at 01:15 GMT_

---

Regarding the usefulness of response files, I know long one liners are of course possible, but they quickly become unwieldy. That is one reason why a file that can use multiple lines is convenient. Also, they then never get lost in the bash history the way a command you once typed does, so they are very useful for that purpose as well. I've found them to be very useful on multiple other projects that are executed via command line and it requires very little code. :-) I also already wrote demo code for groups at my work that uses it so the functionality would already be more painful for me to remove than keep...

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Monday Nov 11, 2013 at 01:18 GMT_

---

Alright then, though you could also just save the whole command in your notes/file/email so the user does not even forget which option flags to use. But I don't mind. If you prefer to have the option, then we'll add it. As said, it won't hurt and you've anyways implemented it already.

**Comment by schuhschuh** [comment]
<a href="https://github.com/ahundt"><img src="https://avatars.githubusercontent.com/u/55744?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [ahundt](https://github.com/ahundt)**
_Monday Nov 11, 2013 at 01:19 GMT_

---

Cool, thanks!

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 00:56 GMT_

---

The response file feature has been merged with modifications by commit c3a817f75c0ced996fb1b1d89d768ee3cb0c4165.

**Comment by schuhschuh** [comment]
<a href="https://github.com/schuhschuh"><img src="https://avatars.githubusercontent.com/u/975451?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [schuhschuh](https://github.com/schuhschuh)**
_Wednesday Nov 13, 2013 at 01:52 GMT_

---

I will close this pull request as the changes are already in my repository and should be merged into master soon as well. Next time when I receive a pull request before having pulled your changes already, I will merge it through the GitHub interface.
